### PR TITLE
Correctly resolve dispatch keys for PyOperator

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96306
* #96304
* #96302

Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/_ops.py [hunk 1]
```diff
@@ -218,8 +222,14 @@ def dispatch(self, dispatch_key, *args, **kwargs):
             # TODO(voz): The idea behind this is that we do not yet support dispatch by key + mode, only key.
             return self.python_key_mode_table[type(curr_mode)](*args, **kwargs)
 
-        assert dispatch_key in self.py_kernels, dispatch_key
-        return self.py_kernels[dispatch_key](*args, **kwargs)
+        final_key = resolve_key(self, dispatch_key)
+
+        assert not isinstance(
+            final_key, str
+        ), f"C++ fallbacks don't work for PyOperator, please reimplement {final_key}"
```
Comment (zou3519) [inline torch/_ops.py:229]: I might not be reading this right, but how can `resolve_key` return a str? It seems to only return torch._C.DispatchKey

**Review by zou3519 — APPROVED** [review]
LGTM

File: torch/_ops.py (hunk 1 above)
Comment (ezyang) [reply torch/_ops.py:229]: oops, you're right

**Pull request closed without merging**